Keycloak fails to start when uninstalling custom provider

### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

core

### Describe the bug

Hi.

I'm working with KeyCloak (_both, with a Docker image and a locally downloaded v22.0.5_) and a custom mapper for OIDC Tokens.

Basically, I've used the code in this URL:
https://betterjavacode.com/programming/custom-protocol-mapper-keycloak

I've generated a JAR and put it into the /opt/keycloak/providers folder. Let's call it: "`provider-cpm_1.0.0.jar`".

Later, I've used it (I can see it as a mapper) in a custom scope within a own Realm's client. It works.
( KeyCloak in a docker container from: keycloak/keycloak:latest )

But, if I want to "uninstall" the mapper from KeyCloak, I just delete the file, as par:
https://www.keycloak.org/server/configuration-provider#_installing_and_uninstalling_a_provider

The problem is that, if I stop the container and start it again, then a nice "`NoSuchFileException`" error message comes:

```
2023-11-01 11:50:37 The DelayedHandler was closed before any children handlers were configured. Messages will be written to stderr.
2023-11-01 11:50:37 2023-11-01 10:50:37,162 DEBUG [org.jboss.logging] (main) Logging Provider: org.jboss.logging.JBossLogManagerProvider
2023-11-01 11:50:37 
2023-11-01 11:50:37 Exception in thread "main" java.lang.reflect.InvocationTargetException
2023-11-01 11:50:37     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-11-01 11:50:37     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
2023-11-01 11:50:37     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-11-01 11:50:37     at java.base/java.lang.reflect.Method.invoke(Method.java:568)
2023-11-01 11:50:37     at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
2023-11-01 11:50:37     at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
2023-11-01 11:50:37 Caused by: java.lang.RuntimeException: Failed to open /opt/keycloak/lib/../providers/provider-cpm_1.0.0.jar
2023-11-01 11:50:37     at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:172)
2023-11-01 11:50:37     at io.quarkus.bootstrap.runner.JarResource.readLockAcquireAndGetJarReference(JarResource.java:153)
2023-11-01 11:50:37     at io.quarkus.bootstrap.runner.JarResource.getResourceURL(JarResource.java:101)
2023-11-01 11:50:37     at io.quarkus.bootstrap.runner.RunnerClassLoader.findResources(RunnerClassLoader.java:214)
2023-11-01 11:50:37     at java.base/java.lang.ClassLoader.getResources(ClassLoader.java:1473)
2023-11-01 11:50:37     at io.smallrye.common.classloader.ClassPathUtils.consumeAsPaths(ClassPathUtils.java:84)
2023-11-01 11:50:37     at io.smallrye.config.AbstractLocationConfigSourceLoader.tryClassPath(AbstractLocationConfigSourceLoader.java:139)
2023-11-01 11:50:37     at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:102)
2023-11-01 11:50:37     at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:85)
2023-11-01 11:50:37     at io.quarkus.runtime.configuration.ApplicationPropertiesConfigSourceLoader$InClassPath.getConfigSources(ApplicationPropertiesConfigSourceLoader.java:30)
2023-11-01 11:50:37     at io.quarkus.runtime.configuration.ApplicationPropertiesConfigSourceLoader$InClassPath.getConfigSources(ApplicationPropertiesConfigSourceLoader.java:27)
2023-11-01 11:50:37     at io.smallrye.config.SmallRyeConfig$ConfigSources.buildSources(SmallRyeConfig.java:581)
2023-11-01 11:50:37     at io.smallrye.config.SmallRyeConfig$ConfigSources.<init>(SmallRyeConfig.java:537)
2023-11-01 11:50:37     at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:69)
2023-11-01 11:50:37     at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:629)
2023-11-01 11:50:37     at io.quarkus.runtime.configuration.QuarkusConfigFactory.getConfigFor(QuarkusConfigFactory.java:29)
2023-11-01 11:50:37     at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:76)
2023-11-01 11:50:37     at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:64)
2023-11-01 11:50:37     at org.keycloak.quarkus.runtime.configuration.Configuration.getConfig(Configuration.java:51)
2023-11-01 11:50:37     at org.keycloak.quarkus.runtime.cli.Picocli.hasConfigChanges(Picocli.java:239)
2023-11-01 11:50:37     at org.keycloak.quarkus.runtime.cli.Picocli.requiresReAugmentation(Picocli.java:141)
2023-11-01 11:50:37     at org.keycloak.quarkus.runtime.cli.Picocli.runReAugmentationIfNeeded(Picocli.java:126)
2023-11-01 11:50:37     at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:95)
2023-11-01 11:50:37     at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:88)
2023-11-01 11:50:37     ... 6 more
2023-11-01 11:50:37 Caused by: java.nio.file.NoSuchFileException: /opt/keycloak/lib/../providers/provider-cpm_1.0.0.jar
2023-11-01 11:50:37     at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
2023-11-01 11:50:37     at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
2023-11-01 11:50:37     at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
2023-11-01 11:50:37     at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
2023-11-01 11:50:37     at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:148)
2023-11-01 11:50:37     at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
2023-11-01 11:50:37     at java.base/java.nio.file.Files.readAttributes(Files.java:1851)
2023-11-01 11:50:37     at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1394)
2023-11-01 11:50:37     at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:716)
2023-11-01 11:50:37     at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:250)
2023-11-01 11:50:37     at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:179)
2023-11-01 11:50:37     at java.base/java.util.jar.JarFile.<init>(JarFile.java:346)
2023-11-01 11:50:37     at io.smallrye.common.io.jar.JarFiles.create(JarFiles.java:33)
2023-11-01 11:50:37     at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:170)
2023-11-01 11:50:37     ... 29 more
```

Any idea of how to avoid this situation?

Many thanks in advance.

### Version

22.0.5

### Expected behavior

The "`build`" step, just before "`start | start-dev`" should take care of "non-existing" providers and do not stop server to initiate.
Maybe, throwing an error message in the log file, but not stop it; or providing an alternate way to start or instructions on how to remediate the bug.

### Actual behavior

Breaks the startup process, and the eserver is not starting

### How to Reproduce?

1. Put a JAR file for a mapper in the "`/providers`" folder.
2. Start the server
3. Remove the file
4. Stop the server
5. Restart again

### Anything else?

_No response_

I just tested this again on 23.0.3. I removed a provider jar file from the `providers` directory and started Keycloak with `start-dev`, triggering a rebuild. It started with an Exception (see below) but did finish successfully and Keycloak remained online afterwards. So that is an improvement. It would be nice if the exception was handled gracefully, e.g. by logging something like "Provider xxx no longer present, removing it."

```
java.lang.RuntimeException: Failed to open REDACTED_PATH_TO_JAR
        at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:172)
        at io.quarkus.bootstrap.runner.JarResource.readLockAcquireAndGetJarReference(JarResource.java:153)
        at io.quarkus.bootstrap.runner.JarResource.getResourceURL(JarResource.java:101)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.findResource(RunnerClassLoader.java:162)
        at java.base/java.lang.ClassLoader.getResource(ClassLoader.java:1408)
        at java.base/java.lang.ClassLoader.getResourceAsStream(ClassLoader.java:1738)
        at org.codehaus.groovy.reflection.GeneratedMetaMethod$DgmMethodRecord.loadDgmInfo(GeneratedMetaMethod.java:191)
        at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.registerMethods(MetaClassRegistryImpl.java:207)
        at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.<init>(MetaClassRegistryImpl.java:116)
        at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.<init>(MetaClassRegistryImpl.java:94)
        at groovy.lang.GroovySystem.<clinit>(GroovySystem.java:37)
        at org.codehaus.groovy.runtime.InvokerHelper.<clinit>(InvokerHelper.java:71)
        at groovy.lang.GroovyObjectSupport.getDefaultMetaClass(GroovyObjectSupport.java:46)
        at groovy.lang.GroovyObjectSupport.<init>(GroovyObjectSupport.java:32)
        at groovy.lang.Closure.<init>(Closure.java:215)
        at groovy.lang.Closure.<init>(Closure.java:232)
        at groovy.lang.Closure$1.<init>(Closure.java:197)
        at groovy.lang.Closure.<clinit>(Closure.java:197)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:375)
        at picocli.CommandLine$DefaultFactory.loadClosureClass(CommandLine.java:5619)
        at picocli.CommandLine$DefaultFactory.<clinit>(CommandLine.java:5613)
        at picocli.CommandLine$Model$CommandUserObject.getInstance(CommandLine.java:12263)
        at picocli.CommandLine$Model$CommandUserObject.get(CommandLine.java:12288)
        at picocli.CommandLine$Model$FieldBinding.set(CommandLine.java:12107)
        at picocli.CommandLine$Model$CommandReflection.initFromAnnotatedTypedMembers(CommandLine.java:11975)
        at picocli.CommandLine$Model$CommandReflection.initFromAnnotatedMembers(CommandLine.java:11909)
        at picocli.CommandLine$Model$CommandReflection.extractCommandSpec(CommandLine.java:11842)
        at picocli.CommandLine$Model$CommandSpec.forAnnotatedObject(CommandLine.java:6388)
        at picocli.CommandLine.<init>(CommandLine.java:230)
        at picocli.CommandLine.toCommandLine(CommandLine.java:3631)
        at picocli.CommandLine.access$16700(CommandLine.java:148)
        at picocli.CommandLine$Model$CommandReflection.initSubcommands(CommandLine.java:11874)
        at picocli.CommandLine$Model$CommandReflection.extractCommandSpec(CommandLine.java:11840)
        at picocli.CommandLine$Model$CommandSpec.forAnnotatedObject(CommandLine.java:6388)
        at org.keycloak.quarkus.runtime.cli.Picocli.createCommandLine(Picocli.java:448)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:97)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:107)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.nio.file.NoSuchFileException: REDACTED_PATH_TO_JAR
        at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
        at java.base/sun.nio.fs.WindowsFileAttributeViews$Basic.readAttributes(WindowsFileAttributeViews.java:53)
        at java.base/sun.nio.fs.WindowsFileAttributeViews$Basic.readAttributes(WindowsFileAttributeViews.java:38)
        at java.base/sun.nio.fs.WindowsFileSystemProvider.readAttributes(WindowsFileSystemProvider.java:199)
        at java.base/java.nio.file.Files.readAttributes(Files.java:1851)
        at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1428)
        at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:718)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:252)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:181)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:346)
        at io.smallrye.common.io.jar.JarFiles.create(JarFiles.java:33)
        at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:170)
        ... 43 more
```

Hi, @shawkins.

I hope You can "read" this PowerShell script... and it works! (*adapt it as you wish*)
Attached, I add the source code I've been using: "Custom Mappers.zip"
[Custom Mappers.zip](https://github.com/keycloak/keycloak/files/14292989/Custom.Mappers.zip)

I have **TWO** questions:
+ Why removing the custom provider's JAR file stops the server startup?, and
+ Why the server admits two (*or more*) custom providers, with the same [package.class] at a time?


The script with my steps.

```powershell
# ================================================================
# Launch a PowerShell console
# CONSOLE["Test KeyCloak Console"]

# Configurar la consola.
Clear-Host;

# Console's title
$Host.UI.RawUI.WindowTitle = "Test KeyCloak Console";

# Some variables
$ROOT_FOLDER = "C:\Temp\Test KeyCloak";
$KEYCLOAK_VERSION = "23.0.6";
$KEYCLOAK_HOME = "keycloak-$($KEYCLOAK_VERSION)";

# Operations
Set-Location "${ROOT_FOLDER}"


# Download KeyCloak "latest" version
# Check: https://github.com/keycloak/keycloak/releases/

# if ( !( Test-Path ${ROOT_FOLDER} -PathType Container ) ) {
#     New-Item -ItemType Directory -Force -Path ${ROOT_FOLDER} | Out-Null;
# }


# Descargamos el archivo
Set-Location "${ROOT_FOLDER}"
$ProgressPreference = 'SilentlyContinue'    # Subsequent calls do not display UI.
IWR "https://github.com/keycloak/keycloak/releases/download/$($KEYCLOAK_VERSION)/$($KEYCLOAK_HOME).zip" -OutFile "$($KEYCLOAK_HOME).zip"
$ProgressPreference = 'Continue'            # Subsequent calls do display UI.
# FILE: [${ROOT_FOLDER}\$($KEYCLOAK_HOME).zip] (164.677.937 bytes)

# Descomprimimos el archivo descargado
$ProgressPreference = 'SilentlyContinue'    # Subsequent calls do not display UI.
Expand-Archive -Path "${ROOT_FOLDER}\$($KEYCLOAK_HOME).zip" -DestinationPath "${ROOT_FOLDER}\KeyCloak"
$ProgressPreference = 'Continue'            # Subsequent calls do display UI.
# FOLDER: [${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)]


# ================================================================
# Start KeyCloak
# Launch a new PowerShell console
# CONSOLE["Test KeyCloak Server"]
Clear-Host;

# Console's title
$Host.UI.RawUI.WindowTitle = "Test KeyCloak Server";

# Some variables
$ROOT_FOLDER = "C:\Temp\Test KeyCloak";
$KEYCLOAK_VERSION = "23.0.6";
$KEYCLOAK_HOME = "keycloak-$($KEYCLOAK_VERSION)";

# Operations
Set-Location "${ROOT_FOLDER}"

# Arrancamos el servidor
Set-Location "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\bin"
& .\kc.bat start-dev

# Stop server with ^C.
# (will run after ^C)
Set-Location "${ROOT_FOLDER}"
# ================================================================


# ================================================================
# Open Admin console at: http://localhost:8080/
# CONSOLE["Test KeyCloak Console"]
Clear-Host;
Start-Process "http://localhost:8080/";

# CHECK it works.
# CREATE [admin/admin] account
# LOGIN
# CHECK it works.
# STOP the server in server's console
# ================================================================


# ================================================================
# Create mapper: "Mapper 01"
# CONSOLE["Test KeyCloak Console"]
# Use: https://betterjavacode.com/programming/custom-protocol-mapper-keycloak

# Mapper 01
Set-Location "${ROOT_FOLDER}\Custom Mappers\mapper-01"
Clear-Host;

# Maven: Create custom mapper file
mvn clean package;

# Copy custom mapper file
Copy-Item -Path "${ROOT_FOLDER}\Custom Mappers\mapper-01\target\pipogit-cpm-01.jar" -Destination "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\providers";

# List custom mappers
Get-ChildItem "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\providers\*.jar";

# CHECK "pipogit-cpm-01.jar" is listed.

# Back to HOME
Set-Location "${ROOT_FOLDER}"
# ================================================================


# ================================================================
# Launch KeyCloak Server again
# CONSOLE["Test KeyCloak Server"]
# Repeat previous steps
Clear-Host;

# Operations
Set-Location "${ROOT_FOLDER}"

# Arrancamos el servidor
Set-Location "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\bin"
& .\kc.bat start-dev

# Stop server with ^C.
# (will run after ^C)
Set-Location "${ROOT_FOLDER}"
# ================================================================


# ================================================================
# Open Admin console at: http://localhost:8080/
# CONSOLE["Test KeyCloak Console"]
Clear-Host;
Start-Process "http://localhost:8080/";
# Login
# GOTO: Client Scopes
# Select one (i.e.: "address")
# GOTO: "Mappers"
# Select: "Add mapper" > "By configuration"
# CHECK that "PipoGIT Custom Mapper 01" exists.
# STOP the server in server's console
# ================================================================



# ================================================================
# Create mapper: "Mapper 02"
# CONSOLE["Test KeyCloak Console"]
# Use: https://betterjavacode.com/programming/custom-protocol-mapper-keycloak

# Mapper 02
Set-Location "${ROOT_FOLDER}\Custom Mappers\mapper-02"
Clear-Host;

# Maven: Create custom mapper file
mvn clean package;

# Copy custom mapper file
Copy-Item -Path "${ROOT_FOLDER}\Custom Mappers\mapper-02\target\pipogit-cpm-02.jar" -Destination "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\providers";

# List custom mappers
Get-ChildItem "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\providers\*.jar";

# CHECK "pipogit-cpm-01.jar" AND "pipogit-cpm-02.jar" are listed.

# Back to HOME
Set-Location "${ROOT_FOLDER}"
# ================================================================


# ================================================================
# Launch KeyCloak Server again
# CONSOLE["Test KeyCloak Server"]
# Repeat previous steps
Clear-Host;

# Operations
Set-Location "${ROOT_FOLDER}"

# Arrancamos el servidor
Set-Location "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\bin"
& .\kc.bat start-dev

# Stop server with ^C.
Set-Location "${ROOT_FOLDER}"
# ================================================================


# ================================================================
# Open Admin console at: http://localhost:8080/
# CONSOLE["Test KeyCloak Console"]
Clear-Host;
Start-Process "http://localhost:8080/";
# Login
# GOTO: Client Scopes
# Select one (i.e.: "address")
# GOTO: "Mappers"
# Select: "Add mapper" > "By configuration"
# CHECK that "PipoGIT Custom Mapper 01" exists.
# CHECK that "PipoGIT Custom Mapper 02" DOES NOT exists.
# STOP the server in server's console
# ================================================================


# ================================================================
#
# *** FINAL TEST ***
#
# CONSOLE["Test KeyCloak Console"]
Clear-Host;

# List custom mappers
Get-ChildItem "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\providers\*.jar";

# Delete "Custom Mapper 02" from "pipogit-cpm-02.jar"
Remove-Item "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\providers\pipogit-cpm-02.jar";

# List custom mappers
Get-ChildItem "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\providers\*.jar";
# ================================================================


# ================================================================
# Launch KeyCloak Server again
# CONSOLE["Test KeyCloak Server"]
# Repeat previous steps
Clear-Host;

# Operations
Set-Location "${ROOT_FOLDER}"

# Arrancamos el servidor
Set-Location "${ROOT_FOLDER}\KeyCloak\$($KEYCLOAK_HOME)\bin"
& .\kc.bat start-dev

# CHECK Error Log Messages:
<#
The DelayedHandler was closed before any children handlers were configured. Messages will be written to stderr.
2024-02-15 08:44:36,341 DEBUG [org.jboss.logging] (main) Logging Provider: org.jboss.logging.JBossLogManagerProvider

Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.lang.RuntimeException: Failed to open C:\Temp\Test KeyCloak\KeyCloak\keycloak-23.0.6\lib\..\providers\pipogit-cpm-02.jar
        at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:172)
        at io.quarkus.bootstrap.runner.JarResource.readLockAcquireAndGetJarReference(JarResource.java:153)
        at io.quarkus.bootstrap.runner.JarResource.getResourceURL(JarResource.java:101)
        at io.quarkus.bootstrap.runner.RunnerClassLoader.findResources(RunnerClassLoader.java:214)
        at java.base/java.lang.ClassLoader.getResources(ClassLoader.java:1469)
        at io.smallrye.common.classloader.ClassPathUtils.consumeAsPaths(ClassPathUtils.java:84)
        at io.smallrye.config.AbstractLocationConfigSourceLoader.tryClassPath(AbstractLocationConfigSourceLoader.java:139)
        at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:102)
        at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:85)
        at io.quarkus.runtime.configuration.ApplicationPropertiesConfigSourceLoader$InClassPath.getConfigSources(ApplicationPropertiesConfigSourceLoader.java:30)
        at io.quarkus.runtime.configuration.ApplicationPropertiesConfigSourceLoader$InClassPath.getConfigSources(ApplicationPropertiesConfigSourceLoader.java:27)
        at io.smallrye.config.SmallRyeConfig$ConfigSources.buildSources(SmallRyeConfig.java:581)
        at io.smallrye.config.SmallRyeConfig$ConfigSources.<init>(SmallRyeConfig.java:537)
        at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:69)
        at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:629)
        at io.quarkus.runtime.configuration.QuarkusConfigFactory.getConfigFor(QuarkusConfigFactory.java:29)
        at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:76)
        at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:64)
        at org.keycloak.quarkus.runtime.configuration.Configuration.getConfig(Configuration.java:51)
        at org.keycloak.quarkus.runtime.cli.Picocli.hasConfigChanges(Picocli.java:335)
        at org.keycloak.quarkus.runtime.cli.Picocli.requiresReAugmentation(Picocli.java:169)
        at org.keycloak.quarkus.runtime.cli.Picocli.runReAugmentationIfNeeded(Picocli.java:153)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:105)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:107)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        ... 3 more
Caused by: java.nio.file.NoSuchFileException: C:\Temp\Test KeyCloak\KeyCloak\keycloak-23.0.6\lib\..\providers\pipogit-cpm-02.jar
        at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
        at java.base/sun.nio.fs.WindowsFileAttributeViews$Basic.readAttributes(WindowsFileAttributeViews.java:53)
        at java.base/sun.nio.fs.WindowsFileAttributeViews$Basic.readAttributes(WindowsFileAttributeViews.java:38)
        at java.base/sun.nio.fs.WindowsFileSystemProvider.readAttributes(WindowsFileSystemProvider.java:197)
        at java.base/java.nio.file.Files.readAttributes(Files.java:1848)
        at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1414)
        at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:723)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:250)
        at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:179)
        at java.base/java.util.jar.JarFile.<init>(JarFile.java:339)
        at io.smallrye.common.io.jar.JarFiles.create(JarFiles.java:33)
        at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:170)
        ... 27 more
#>
# ================================================================


```

--- a/docs/documentation/server_development/topics/providers.adoc
+++ b/docs/documentation/server_development/topics/providers.adoc
@@ -73,7 +73,7 @@ public class MyThemeSelectorProvider implements ThemeSelectorProvider {
     }
 
     @Override
-	public void close() {
+    public void close() {
     }
 }
 ----
@@ -216,11 +216,25 @@ need to be retrieved with the usage of `KeycloakSessionFactory`. It is not recom
 
 === Registering provider implementations
 
-Providers are registered with the server by simply copying them to the `providers` directory.
+Providers are registered with the server by simply copying the JAR file to the `providers` directory.
 
 If your provider needs additional dependencies not already provided by Keycloak copy these to the `providers` directory.
 
-After registering new providers or dependencies Keycloak needs to be re-built with the `kc.[sh|bat] build` command.
+[NOTE]
+====
+Provider JARs are not loaded in isolated classloaders, so do not include resources or classes in your provider JARs that conflict with built-in resources or classes. 
+In particular the inclusion of an application.properties file will cause auto-build to fail if the provider JAR is removed. 
+If you have included conflicting classes, you will see a split package warning in the start log for the server. That should be resolved by removing or repackaging the offending classes. 
+However there is no warning if you have conflicting resource files. You should either ensure that your JAR's resource files have path names that contain something unique to that provider, 
+or you can check for the existence of `some.file` in the JAR contents under the `"install root"/lib/lib/main` directory with something like:
+ 
+[source,bash]
+----
+find . -type f -name "*.jar" -exec unzip -l {} \; | grep some.file
+----
+====    
+
+After registering new providers or dependencies Keycloak needs to be re-built with a non-optimized start or the `kc.[sh|bat] build` command.
 
 ==== Disabling a provider
 

